### ISSUE TITLE
When fixing the stream, also need to fix Category and Playlists

### ISSUE THREAD

**Issue by tfabris** [tfabris/CrowCam#75]
There is existing code in CrowCam.sh for fixing the following things if they are changed:
- Visibility (public, unlisted, private)
- Stream Key

If the Stream Key changes, that's actually a situation where a whole new broadcast had to be created. Some of the settings are lost in that case. Most of the stuff is copied over from your last livestream (luckily things like the text description seem to be kept), but I have noticed that the following things will change each time, and need to be fixed like the Visibility is currently getting fixed:
- Category (i.e., "Pets and Animals")
   - Note that Category is not the same as "Tags" (luckily that seems to be kept).
   - ~~See later comment: I found the general setting for Category so it doesn't need to be automated.~~ Reinstated - this still needs to be fixed.
- Playlists (i.e., "CrowCam Archives")
   - This is the playlist that the videos automatically go into when they are done recording.

**Comment by tfabris** [comment]
Update: I was able to find a setting in the default channel settings. Go to https://studio.youtube.com/ and press SETTINGS in the lower left corner. Press UPLOAD DEFAULTS and then press ADVANCED SETTINGS. This gets you to the default setting for "Category".  With this configured I don't think we need to automate the "Category", I believe it's a set-once-and-forget setting.

Still investigating "Playlists".

**Comment by tfabris** [comment]
Status:
- In the VIDEO resource (whose ID happens to be the same as the "broadcastID" that I retrieve from the new broadcast when I create it), there is a "categoryId": "24" entry in the "snippet" section. 24 is the default "Entertainment" category, 15 is the "Pets & Animals" category. [Docs are here](https://developers.google.com/youtube/v3/docs/videos#snippet.categoryId). See if we can update this after we create a new video and the stream has gone live.

To get the list of all possible category IDs in your country, you have to use the [VidoeCategories: list function](https://developers.google.com/youtube/v3/docs/videoCategories/list), with the parameters `part=snippet&regionCode=US` where the "US" part is your desired region code from the standard two-letter region codes like the ones [listed here](https://www.iban.com/country-codes).

Here is the list for the "US" regionCode:
1 Film & Animation
2 Autos & Vehicles
10 Music
15 Pets & Animals
17 Sports
18 Short Movies
19 Travel & Events
20 Gaming
21 Videoblogging
22 People & Blogs
23 Comedy
24 Entertainment
25 News & Politics
26 Howto & Style
27 Education
28 Science & Technology
29 Nonprofits & Activism
30 Movies
31 Anime/Animation
32 Action/Adventure
33 Classics
34 Comedy
35 Documentary
36 Drama
37 Family
38 Foreign
39 Horror
40 Sci-Fi/Fantasy
41 Thriller
42 Shorts
43 Shows
44 Trailers

- There are no "playlist" related items in either the broadcast, livestream, or video resources. My guess is that if I want the video to be automatically added to a playlist, I need to open the desired playlist when I create the video stream and add it there.

To do:
- See if I can do both of those things at the tail end of the CreateNewStream function. After the video is created, add it to the top of the playlist and then update the categoryId of the video's "snippet".

**Comment by tfabris** [comment]
Updating the categoryId from the API explorer works, now I just need to implement it in code. You must use "part=snippet" and the BODY DATA of the request (not the parameters) contains the ID of the video that you want to update. You must update both the title and the cateogryID, those happen to be the minimum things. Also it must be a PUT operation. Like this:

```
PUT https://youtube.googleapis.com/youtube/v3/videos?part=snippet

{
"id": "icdhnUth9XY"
  "snippet":
  {
    "title": "CrowCam",
    "categoryId": "15"
  }
}
```

**Comment by tfabris** [comment]
Adding to the playlist works (I have to fix the ordering position, it oddly doesn't default to the top), but writing the "snippet" with just those two items, it CLOBBERS every other item in the snipped, like Description and Tags. ARGH! Need to fix that before I can put this code into action.

**Comment by tfabris** [comment]
Fixed with checkin [dac2b13](https://github.com/tfabris/CrowCam/commit/dac2b13a1f3664e6d6b5246228c85d0c0cbbf415) - Youtube has a bug where when you fix the categoryId by sending a "Snippet" to the video, even if your snippet is only populated with the minimum fields (Title and CategoryID), it clobbers the entire rest of the snippet! That means that Description, Tags, and Thumbnail are overwritten. This is awful. Work around this issue by adding default Description, Tags, and Thumbnail directly into the sources and write all that stuff to the video, including uploading a new Thumbnail JPG every fscking time. Wow YouTube makes it so hard.

Adding to the playlist works as expected: You have to open up the playlist and then add the video to the playlist (not the other way around), and you must add it at position: 0 in order to float it to the top. This incidentally also fixes bug #78 which is a nice bonus.
